src/main/java/uk/gov/justice/hmpps/prison/service/OffenderTransactionHistoryService.java

```diff
@@ -81,30 +83,26 @@ private void validate(final String offenderNo, final String accountCode, final L
     }
 
     private List<OffenderTransactionHistory> getAllTransactionsWithRunningBalance(final String offenderNo) {
-        final var allOffenderTransactions = historyRepository
+        final var transactions = historyRepository
             .findByOffender_NomsId(offenderNo)
             .stream()
             .sorted(SORT_BY_OLDEST_DATE)
-            .collect(groupingBy(OffenderTransactionHistory::getAgencyId));
+            .collect(groupingBy(transaction -> Pair.of(transaction.getAgencyId(), transaction.getAccountType())));
 
-        allOffenderTransactions.keySet().forEach(agencyId -> {
+        return transactions.keySet().stream().flatMap(agencyAccountType -> {
```

Comment:
Would it be useful to change the variable name for agencyAccountType, given that it is now a Pair?